### PR TITLE
Extracted config ordering logic into a helper function

### PR DESCRIPTION
**Pull request by melinath**
Related to https://github.com/hashicorp/terraform-provider-google/issues/16243 - if this looks good I'll do a follow-up PR to fix that & related problems.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: mmv1/third_party/terraform/tpgresource/utils_test.go [hunk 1]
```diff
@@ -37,6 +37,145 @@ var fictionalSchema = map[string]*schema.Schema{
...
+			configData: []string{},
+			apiData:    []string{},
+			want:       []string{},
+		},
+		"config data with empty api data": {
+			configData: []string{"one", "two"},
+			apiData:    []string{},
+			want:       []string{},
+		},
+		"empty config data with api data": {
```
Comment (askubis) [inline mmv1/third_party/terraform/tpgresource/utils_test.go:56]: Nit: why does the order change here?
I know this does not affect the results.

**Review by askubis — APPROVED** [review]
LGTM, 2 nits/questions

File: mmv1/third_party/terraform/tpgresource/utils_test.go [hunk 2]
```diff
@@ -37,6 +37,145 @@ var fictionalSchema = map[string]*schema.Schema{
...
+				t.Fatalf("Wanted error, got none")
+			}
+			if !tc.wantError && sorted != tc.want {
+				t.Fatalf("sorted result is incorrect. want %v, got %v", sorted, tc.want)
+			}
+		})
+	}
+}
+
+func TestSortMapsByConfigOrder(t *testing.T) {
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/tpgresource/utils_test.go:134]: There are a couple of mismatched variable types here that I spotted when trying to run the test locally

File: mmv1/third_party/terraform/tpgresource/utils_test.go [hunk 3]
```diff
@@ -37,6 +37,145 @@ var fictionalSchema = map[string]*schema.Schema{
...
+			idKey:     "bar",
+			wantError: true,
+		},
+	}
+
+	for tn, tc := range cases {
+		tc := tc
+		t.Run(tn, func(t *testing.T) {
+			t.Parallel()
+			sorted, err := SortMapsByConfigOrder(tc.configData, tc.apiData, tc.idKey)
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/tpgresource/utils_test.go:164]:
```suggestion
			sorted, err := tpgresource.SortMapsByConfigOrder(tc.configData, tc.apiData, tc.idKey)
```

I think the tests all need to be updated in a similar way to access the new functions

File: mmv1/third_party/terraform/tpgresource/utils.go [hunk 4]
```diff
@@ -237,6 +238,87 @@ func ExpandStringMap(d TerraformResourceData, key string) map[string]string {
...
+			return nil, fmt.Errorf("configData element at %d has duplicate value `%s`", index, item)
+		}
+		configOrder[item] = index
+	}
+
+	apiSeen := map[string]struct{}{}
+	byConfigIndex := map[int]string{}
+	newElements := []string{}
+	for index, item := range apiData {
+		_, ok := apiSeen[item]
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/tpgresource/utils.go:258]: I can't see anywhere where `apiSeen` has key:value pairs added to it - should this be happening at the end of this for loop block?

**Review by SarahFrench — COMMENTED** [review]
This looks good to me so far, but I've left a question about `apiSeen` and currently there are a few problems preventing running unit tests

File: mmv1/third_party/terraform/go.mod.erb [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <% autogen_exception -%>
-module github.com/hashicorp/terraform-provider-google
+module github.com/hashicorp/terraform-provider-google-beta
```
[the remark this answers is not in the document]
Comment (melinath) [reply mmv1/third_party/terraform/go.mod.erb:2]: yep, it's a copy-paste error.

[the remark this answers is not in the document]
Comment (melinath) [reply mmv1/third_party/terraform/go.mod.erb:2]: fixed in 6afc8e6be

File: mmv1/third_party/terraform/tpgresource/utils_test.go (hunk 1 above)
Comment (melinath) [reply mmv1/third_party/terraform/tpgresource/utils_test.go:56]: Currently, we sort the new items alphabetically. It's not strictly necessary but may make things easier to read. (It also matches the current instance group manager logic.)

File: mmv1/third_party/terraform/tpgresource/utils_test.go [hunk 6]
```diff
@@ -37,6 +37,145 @@ var fictionalSchema = map[string]*schema.Schema{
...
+			if err != nil {
+				if !tc.wantError {
+					t.Fatalf("Unexpected error: %s", err)
+				}
+			} else if tc.wantError {
+				t.Fatalf("Wanted error, got none")
+			}
+			if !tc.wantError && sorted != tc.want {
+				t.Fatalf("sorted result is incorrect. want %v, got %v", sorted, tc.want)
```
[the remark this answers is not in the document]
Comment (melinath) [reply mmv1/third_party/terraform/tpgresource/utils_test.go:173]: Fixed in 546719732

File: mmv1/third_party/terraform/services/compute/resource_compute_instance_group_manager.go.erb [hunk 7]
```diff
@@ -1304,54 +1304,29 @@ func flattenStatefulPolicyStatefulExternalIps(d *schema.ResourceData, statefulPo
...
-
-		index, found := order[interfaceName]
-		if !found {
-			unexpectedIps = append(unexpectedIps, data)
-			continue
-		}
-		orderedResult[index] = data // Put elements from API response in order that matches the config
+		apiData = append(apiData, data)
 	}
-	sort.Slice(unexpectedIps, func(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (melinath) [reply mmv1/third_party/terraform/services/compute/resource_compute_instance_group_manager.go.erb:1337]: Fixed in cdb87ba06

File: mmv1/third_party/terraform/tpgresource/utils_test.go (hunk 2 above)
Comment (melinath) [reply mmv1/third_party/terraform/tpgresource/utils_test.go:134]: fixed in 0e3ccd403

File: mmv1/third_party/terraform/tpgresource/utils_test.go (hunk 3 above)
Comment (melinath) [reply mmv1/third_party/terraform/tpgresource/utils_test.go:164]: Fixed in 0e3ccd403

File: mmv1/third_party/terraform/tpgresource/utils.go (hunk 4 above)
Comment (melinath) [reply mmv1/third_party/terraform/tpgresource/utils.go:258]: Fixed in 0e3ccd403

**Comment by melinath** [comment]
`TestAccSqlUser_postgresIAM` failure is unrelated.

**Review by SarahFrench — APPROVED** [review]
This looks good to me - the test cases look good and it's reassuring that compute/resource_compute_instance_group_manager.go.erb could be refactored and the test for its flattener function is passing still. Looking forward to seeing this applied to that other issue you linked!

File: mmv1/third_party/terraform/tpgresource/utils_test.go (hunk 1 above)
Comment (askubis) [reply mmv1/third_party/terraform/tpgresource/utils_test.go:56]: OK, Understood. LGTM

**Review by askubis — APPROVED** [review]
LGTM & Approval

**Pull request merged**